security/oc_keypair.c

```diff
@@ -44,86 +47,135 @@ oc_sec_get_ecdsa_keypair(size_t device)
   return NULL;
 }
 
+size_t
+oc_sec_ecdsa_count_keypairs()
+{
+  return (size_t)oc_list_length(g_oc_keypairs);
+}
+
 void
-oc_free_ecdsa_keypairs(void)
+oc_sec_ecdsa_free_keypairs(void)
 {
-  oc_ecdsa_keypair_t *kp = (oc_ecdsa_keypair_t *)oc_list_pop(oc_keypairs);
+  oc_ecdsa_keypair_t *kp = (oc_ecdsa_keypair_t *)oc_list_pop(g_oc_keypairs);
   while (kp) {
-    oc_memb_free(&oc_keypairs_s, kp);
-    kp = (oc_ecdsa_keypair_t *)oc_list_pop(oc_keypairs);
+    oc_memb_free(&g_oc_keypairs_s, kp);
+    kp = (oc_ecdsa_keypair_t *)oc_list_pop(g_oc_keypairs);
   }
 }
 
 static oc_ecdsa_keypair_t *
-add_keypair(size_t device)
+ecdsa_add_keypair(size_t device)
 {
-  oc_ecdsa_keypair_t *kp = oc_memb_alloc(&oc_keypairs_s);
-  if (!kp) {
+  oc_ecdsa_keypair_t *kp = oc_memb_alloc(&g_oc_keypairs_s);
+  if (kp == NULL) {
+    OC_ERR("cannot allocate keypair");
     return NULL;
   }
   kp->device = device;
-  oc_list_add(oc_keypairs, kp);
+  oc_list_add(g_oc_keypairs, kp);
   return kp;
 }
 
 bool
-oc_sec_decode_ecdsa_keypair(oc_rep_t *rep, size_t device)
+oc_sec_ecdsa_decode_keypair(const oc_rep_t *rep, oc_ecdsa_keypair_t *kp)
 {
-  oc_ecdsa_keypair_t *kp = oc_sec_get_ecdsa_keypair(device);
-  if (!kp) {
-    kp = add_keypair(device);
-    if (!kp) {
-      return false;
+#define PROP_PUBLIC_KEY "public_key"
+#define PROP_PRIVATE_KEY "private_key"
+#define STR_LEN(x) (sizeof(x) - 1)
+
+  const uint8_t *public_key = NULL;
+  const uint8_t *private_key = NULL;
+  uint8_t private_key_size = 0;
+  uint8_t public_key_size = 0;
```

Comment:
for p521 it is big enough?

Reply:
maybe `size_t` will be better.

Reply:
With `MBEDTLS_ECP_DP_SECP521R1_ENABLED` defined -> private key has size=223 and public key has size=158. So it would fit, but it's assigned to a `size_t` member of the keypair struct, so no there's no reason not use `size_t` anyway.